apis/core.oam.dev/v1beta1/core_types.go

```diff
@@ -160,6 +160,9 @@ type TraitDefinitionSpec struct {
 	// ControlPlaneOnly defines which cluster is dispatched to
 	// +optional
 	ControlPlaneOnly bool `json:"controlPlaneOnly,omitempty"`
+	// PreApply represents dispatch manifests in advance
+	// +optional
+	PreApply bool `json:"preApply,omitempty"`
```

Comment:
we need a better name

Reply:
How about `ApplyPriority` and let it be a number. Default to be 0. Higher number indicates earlier apply.

Reply:
I doesn't work. A number is a fixed value only work on fixed comparison, while definition itself is a dynamic module can be used in any composition.

Reply:
I think it should just control which needs to be dispatch in advance and make sure it's healthy. If the concept of priority is introduced, the complexity will increase greatly. New features require knowledge of all other priorities.

Reply:
> A number is a fixed value only work on fixed comparison, while definition itself is a dynamic module can be used in any composition.

But how could you describe a dynamic comparison? At least `ApplyPriority` is a superset for `preApply`. Although I agree it has limitations.

> If the concept of priority is introduced, the complexity will increase greatly.

Yes. I agree with it. The implementation of priority will be more complex. But then we need to make it clear what this PR wants to solve and the boundary here.

For example, in current KubeVela application, even if without the PR here, user can still use Workflow / DependsOn to make resource dispatch in order. Therefore, AFAIS, this PR is not a necessity for features (The ConfigMap case could be split into `creating ConfigMap` and `mounting ConfigMap`), as discussed in previous PR. One of the big progress here is that we are moving the responsibility for resource dispatch order controlling from end-user to system-operators (who design traits and components). Then I believe it will be okay for asking system-operators to gain `the knowledge of all other priorities`.

When you want to `make sure it's healthy`, this can be difficult in KubeVela model since you need `healthPolicy` in definitions. Therefore, you have lines of codes for the implementation for this feature in this PR, which I think although useful but might be better if it can be generalized into more scenarios.

Anyway, `Priority` is just a recommendation implementation from my perspective. The code here generally LGTM. Thanks for the contribution.